### PR TITLE
Suppress stack trace in test

### PR DESCRIPTION
**Pull request by som-snytt**

### CODE REVIEW

File: src/testkit/scala/tools/testkit/AssertUtil.scala [hunk 1]
```diff
@@ -201,9 +201,41 @@ object AssertUtil {
 
   /** Like Await.ready but return false on timeout, true on completion, throw InterruptedException. */
   def readyOrNot(awaitable: Awaitable[_]): Boolean = Try(Await.ready(awaitable, TestDuration.Standard)).isSuccess
+
+  /** Syntax to clarify which is the expected value. */
+  def assertEqualTo[A](expected: A)(actual: A): Unit = assertEquals(expected, actual)
```
Comment (hrhino) [inline src/testkit/scala/tools/testkit/AssertUtil.scala]: This cries for a shiny new extension method: `(1 + 1) assertEqualTo 2`...

File: src/testkit/scala/tools/testkit/AssertUtil.scala [hunk 2]
```diff
@@ -201,9 +201,41 @@ object AssertUtil {
...
+  def run(): Unit = {
+    val group = new ThreadGroup("stackless") {
+      override def uncaughtException(t: Thread, e: Throwable): Unit = addError(t, e)
+    }
+    val thread = new Thread(group, () => { result = Some(body) ; () })
+    thread.start()
+    thread.join()
+  }
+
+  private def addError(t: Thread, e: Throwable): Unit = synchronized(errors.+=((t, e)))
```
Comment (hrhino) [inline src/testkit/scala/tools/testkit/AssertUtil.scala]: what do you have against infix?

File: test/junit/scala/sys/process/ProcessTest.scala [hunk 3]
```diff
@@ -102,14 +103,23 @@ class ProcessTest {
...
-    } finally {
-      Files.delete(out)
+        assertEquals(1, p.!)
+        assertFalse(failed.get)
+      } finally {
+        Files.delete(out)
+      }
+    }
+    stackless.run()
+    stackless.assert()
```
Comment (hrhino) [inline test/junit/scala/sys/process/ProcessTest.scala]: I'd have made this part of `.run()`, but this is more separation-of-concerns-y.

**Review by hrhino — APPROVED** [review]
Testing PRs never get any love.

File: src/testkit/scala/tools/testkit/AssertUtil.scala (hunk 1 above)
Comment (som-snytt) [reply src/testkit/scala/tools/testkit/AssertUtil.scala]: Don't make me dig out that meme with the red dress.

We recently re-watched The Matrix, which I don't adore as some do, but the daughter said we watched it already, and then oh that's where that meme comes from, meaning the red/blue pill. I can never remember which pill is which. Anyway, the red dress from Matrix is a thing but I think not a meme? This red dress is the distraction that leads to destruction. It was conceived as a test, but the coder guy says he has recycled it for other temptations.

File: src/testkit/scala/tools/testkit/AssertUtil.scala (hunk 2 above)
Comment (som-snytt) [reply src/testkit/scala/tools/testkit/AssertUtil.scala]: Honestly, I always forget whether infix will work or it takes tuples or multiargs wrong.

Comment (dwijnand) [reply src/testkit/scala/tools/testkit/AssertUtil.scala]: Double-up parens to be sure (to be sure)

```scala
errors += ((t, e))
```

File: src/testkit/scala/tools/testkit/AssertUtil.scala [hunk 4]
```diff
@@ -201,9 +201,41 @@ object AssertUtil {
...
+
+  def getResult: Option[A] = result
+
+  def getErrors: List[(Thread, Throwable)] = synchronized(errors.toList)
+
+  def assert(): Unit = getErrors.collect { case (_, e: AssertionError) => e }.headOption.foreach(e => throw e)
```
Comment (dwijnand) [inline src/testkit/scala/tools/testkit/AssertUtil.scala]: Maybe it's best to report them all?

```suggestion
  def assert(): Unit = {
    getErrors.collect { case (_, e: AssertionError) => e } match {
      case Nil      => ()
      case e :: Nil => throw e
      case errors   =>
        val acc = new AssertionError("Multiple assertion errors")
        errors.foldLeft(acc) { (acc, err) =>
          acc.addSuppressed(err)
          acc
        }
        throw acc
    }
  }
```

Or, if you prefer "last exception wins" semantics (like the suppression in `try throw X finally throw Y`):

```suggestion
  def assert(): Unit = {
    getErrors.collect { case (_, e: AssertionError) => e } match {
      case Nil    => ()
      case errors => throw errors.reduceRight { (err, acc) => acc.addSuppressed(err); acc }
    }
  }
```

File: test/junit/scala/sys/process/ProcessTest.scala (hunk 3 above)
Comment (dwijnand) [reply test/junit/scala/sys/process/ProcessTest.scala]: Yeah if the intended usage is to run and assert, I'd make `run` assert, and have a more verbose `runWithoutAssert` path-least-taken option.  (But not a blocker..)

File: src/testkit/scala/tools/testkit/AssertUtil.scala (hunk 1 above)
Comment (dwijnand) [inline src/testkit/scala/tools/testkit/AssertUtil.scala]: Is it also part of the intent to make them not LUB?  And/or is it so the second one can be a block?

**Review by dwijnand — APPROVED** [review]
LGTM too.  A couple of ideas/suggestions.

**Comment by som-snytt** [comment]
I returned some ❤️ and incorporated all the suggestions. I also gave it a better pun, `withoutATrace`.

I'm not sure if it will be hugely useful, but I think I wind up needing something like it on every project. 

Note you don't see suppressed in JUnit 4, but maybe in JUnit 5.

`sys.process` is uniquely perverse in kicking off threads. Usually I'd expect to run tasks on a pool, so that might require a bit more work, to be useful.

I have doubts about `assertEqualsTo` because these methods are always heavily overloaded for specialization, and that's probably not worth it here. I'll probably revert that, or let the maintainer of `sys.process` module do it.

**Comment by lrytz** [comment]
Thank you all!

**Comment by som-snytt** [comment]
The force push included a refactoring, but you know, it's only test code. It will get another review when the build starts failing intermittently.

**Pull request merged**